### PR TITLE
[processor/transform] Add functions for conversion of scalar metric types

### PR DESCRIPTION
**Pull request by BinaryFissionGames**
**Description:**
* Add `convert_gauge_to_sum` and `convert_sum_to_gauge` functions
  * I added in resolution of bool parameters to support the `is_monotonic` parameter, as well.

**Link to tracking Issue:** #10130

**Testing:**
Added some unit tests. Manually tested converting metrics with the processor.

**Documentation:**
Added new functions to docs under a "metrics only" section of the config.

### CODE REVIEW

**Comment by TylerHelmuth** [comment]
I'll do a proper review of this tomorrow, but after a quick overview on my phone it looks promising!

**Comment by TylerHelmuth** [comment]
A fix I made for aggregation_temporality and is_monotonic that changed how they  are accessed to make them usable in conditions was merged in preparation for 0.52.0. It might affect your new functions.

File: processor/transformprocessor/README.md
Comment (TylerHelmuth) [inline processor/transformprocessor/README.md:40]:
```suggestion
Metric only functions:
```

File: processor/transformprocessor/internal/metrics/functions.go
Comment (TylerHelmuth) [inline processor/transformprocessor/internal/metrics/functions.go:58]: From a pdata perspective this will get a metric with the correct data type and leave all the NumberDataPoints intact, but is that all it takes to convert a sum to a gauge?  I want to make sure this is a semantically meaningful conversion.

File: processor/transformprocessor/internal/metrics/functions.go
Comment (TylerHelmuth) [inline processor/transformprocessor/internal/metrics/functions.go:73]: With #10197 getting merged in the accessor for `metric.aggregation_temporality` now [converts to and from the protobuf's numeric definition of aggregation temporality](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/48c784cbe7a96d706fae1f9bd95bfda09eac20c8/processor/transformprocessor/README.md?plain=1#L18).  I know this function doesn't need to use that accessor, but I think we should be consistent.  Either everything should use the protobuf's numeric value or everything should use a string value.  I opted for the protobuf definition since that the guide for other fields, but I am up for arguments advocating for string.

File: processor/transformprocessor/internal/common/parser.go
Comment (TylerHelmuth) [inline processor/transformprocessor/internal/common/parser.go:27]: Thanks for adding bool capabilities!

File: processor/transformprocessor/README.md
Comment (TylerHelmuth) [inline processor/transformprocessor/README.md:43]: Can you add examples below on how to use these functions?  I am curious if bool is handled like `convert_gauge_to_sum("delta", "true")` or like `convert_gauge_to_sum("delta", true)`

File: processor/transformprocessor/factory_test.go
Comment (TylerHelmuth) [inline processor/transformprocessor/factory_test.go:54]: Nice catch

File: processor/transformprocessor/internal/metrics/functions_test.go
Comment (TylerHelmuth) [inline processor/transformprocessor/internal/metrics/functions_test.go:55]: Can you add a "happy path" test case that tests the conversions work successfully?

File: processor/transformprocessor/internal/metrics/functions.go
Comment (BinaryFissionGames) [reply processor/transformprocessor/internal/metrics/functions.go:73]: Yeah, saw this was merged last night.

My argument for string based values here is readability from a user perspective; I wouldn't know without looking at the proto what delta aggregation temporality maps to, for instance. IMO, the string is a lot clearer.

File: processor/transformprocessor/internal/metrics/functions_test.go
Comment (BinaryFissionGames) [reply processor/transformprocessor/internal/metrics/functions_test.go:55]: I added these to separate test functions (`Test_newFunctionCall_Metric_Gauge` & `Test_newFunctionCall_Metric_Sum`) because they don't really fit the mold of the `Test_newFunctionCall_NumberDataPoint` test table (e.g. metric must exist on the transform context and I want to test for both sums and gauges).

File: processor/transformprocessor/README.md
Comment (BinaryFissionGames) [reply processor/transformprocessor/README.md:43]: I'll add one; It works like the latter.

File: processor/transformprocessor/internal/metrics/functions.go
Comment (BinaryFissionGames) [reply processor/transformprocessor/internal/metrics/functions.go:58]: I'll take a close look at this, I see what you mean.

File: processor/transformprocessor/internal/metrics/functions.go
Comment (TylerHelmuth) [reply processor/transformprocessor/internal/metrics/functions.go:73]: Agreed, that readability is much better.  If we switch to using a string for aggregation temporality I would prefer the string usage everywhere, so this switch statement would be moved somewhere it could be used by the accessor as well.  I also wish `pdata` had a string interpretation we could take advantage of instead of making our own.  The current implementation of `.String()` returns the proto's enum name which I wouldn't want to use over the numeric value

```go
var AggregationTemporality_name = map[int32]string{
	0: "AGGREGATION_TEMPORALITY_UNSPECIFIED",
	1: "AGGREGATION_TEMPORALITY_DELTA",
	2: "AGGREGATION_TEMPORALITY_CUMULATIVE",
}
```

@anuraaga @bogdandrutu what do you think?

File: processor/transformprocessor/internal/metrics/functions_test.go
Comment (TylerHelmuth) [reply processor/transformprocessor/internal/metrics/functions_test.go:55]: I'm just blind and missed them lol

Comment (TylerHelmuth) [reply processor/transformprocessor/internal/metrics/functions_test.go:55]: They look good.  Function testing is getting an overall as part of #10101 so I'm good with this implementation for the initial function addtion

File: processor/transformprocessor/internal/metrics/functions.go
Comment (TylerHelmuth) [reply processor/transformprocessor/internal/metrics/functions.go:73]: If we go with a string instead of the number we'll have to document that clearly because the readme and [processing.md doc](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/processing.md#telemetry-query-language) set the expectation that the protobuf/pdata is how fields are accessed/interacted with.

File: processor/transformprocessor/internal/metrics/functions.go
Comment (BinaryFissionGames) [reply processor/transformprocessor/internal/metrics/functions.go:58]: Alright, I took a read of https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/datamodel.md#metric-points

Sums themselves have the two extra fields, which we take into account by having the user specify these fields when converting from gauge.

For the datapoints, the main semantic difference is the timestamps, but I think this is actually "best-effort" conversion as-is.

For a sum, StartTimestamp -> Timestamp give the range of time the sum is computed over.

For a gauge, StartTimestamp is the earliest point the gauge could be recorded, and Timestamp is when the dp was sampled.

I think these timestamps transfer quite well.

* StartTimestamp in a sum being the start of when collection could occur makes sense. You need to know the value at that point in time in order to compute the value of the sum over the whole time range. Semantically, it seems to make sense.

* Timestamp in a sum could technically be any time after start and before "now" (typically, it's the time when the metric was scraped, but it doesn't need to be). Timestamp would represent the end range of the sum; There would necessarily need to be a sample taken at this point in time in order for the sum to be know, so it makes sense to convert this Timestamp to the gauge Timestamp. This one is definitely the most confusing case, but it seems to me like the conversion makes sense.

So, I do think this is a semantically meaningful conversion, even if most of that conversion is basically a no-op.

Comment (BinaryFissionGames) [reply processor/transformprocessor/internal/metrics/functions.go:58]: Yeah, I think the timestamps are logically connected enough that conversion both ways is OK.

Comment (TylerHelmuth) [reply processor/transformprocessor/internal/metrics/functions.go:58]: @open-telemetry/collector-contrib-approvers would also love input from people more versed on otel metrics than I.

Comment (BinaryFissionGames) [reply processor/transformprocessor/internal/metrics/functions.go:58]: Thinking about this more, this creates wrong semantics for Gauge -> Delta Sum (time windows shouldn't overlap). I don't have a good solution right this second, will need to think about it a bit more.

Comment (TylerHelmuth) [reply processor/transformprocessor/internal/metrics/functions.go:58]: monotonicity is something else to thing about.  A gauge can go up and down, so setting `is_monotonic=true` when it goes up and down is not semantically correct, but the gauge could technically have only increased, so sometimes `is_monotonic=true` is semantically correct.

I think overall these 2 functions require immense knowledge about the metrics being converted.  They are definitely "Use at your own risk" types of functions.  Should probably add a disclaimer in the readme.

Comment (BinaryFissionGames) [reply processor/transformprocessor/internal/metrics/functions.go:58]: Yeah, you're right. I think the purpose of these functions is pretty much the source is mis-reporting the metrics actual semantic type. Converting from Sum->Gauge or vice-versa is already breaking the semantics, technically. 

I added a disclaimer to the README for the functions. I think we shouldn't try to be too clever with the conversion, 'cuz these are really "I know what I'm doing" functions.

Comment (TylerHelmuth) [reply processor/transformprocessor/internal/metrics/functions.go:58]: Ya I agree.

Comment (mx-psi) [reply processor/transformprocessor/internal/metrics/functions.go:58]: I don't think we should have functions that are there just for the purpose of fixing something broken down the pipeline and that are not semantically meaningful. Does this have another use case besides fixing a type that was misreported?

Comment (mx-psi) [reply processor/transformprocessor/internal/metrics/functions.go:58]: I see, thanks for the context, I was on vacation last week and didn't make it to the meeting. If this was decided on the SIG I don't think I should block this of course 😄

**Review by TylerHelmuth — APPROVED** [review]
Thanks for adding these functions!

File: processor/transformprocessor/internal/common/parser_test.go
Comment (TylerHelmuth) [inline processor/transformprocessor/internal/common/parser_test.go:354]: The `strp`, `intp`, and `floatp` functions were moved to a `testhelper` package in `common`.  Please update references and move this new helper there as well

Comment (BinaryFissionGames) [reply processor/transformprocessor/internal/common/parser_test.go]: Rebased and incorporated these changes; Will rebase again after #10472 is merged, I had to make those changes to get the tests working.

Comment (TylerHelmuth) [reply processor/transformprocessor/internal/common/parser_test.go]: Unit tests are fixed.

**Comment by TylerHelmuth** [comment]
> Rebased again;
> 
> @TylerHelmuth is this waiting on something before being merged, or do you think it's OK to merge at this point?

Would be good to get 1 more approval I think before read-to-merge.  @anuraaga @bogdandrutu do you want to take a look?

**Review by anuraaga — APPROVED** [review]

**Comment by BinaryFissionGames** [comment]
Thanks @anuraaga and @TylerHelmuth!

@djaglowski could you merge this when you get the chance?

**Pull request merged**
